File: app/src/main/scala/com/trade/validationService/ValidationService.scala
package com.trade.validationService

import com.crankuptheamps.client.{HAClient, Message}
import com.trade.amps.AmpsClientUtil
import com.trade.common.Trade
import play.api.libs.json.Json

object ValidationService extends App {

  val client: HAClient = AmpsClientUtil.connect("ValidationService")
  println("Validation Service connected")

  val ms = client.subscribe("trades.enriched").timeout(0)

  while (true){
    val msg = ms.next();
    val trade = Json.parse(msg.getData).as[Trade]

    if (trade.qty > 0) {
      val json = Json.stringify(Json.toJson(trade))
      client.publish("trades.validated", json)
      println(s"Validated trade: $json")
    } else {
      println(s"Invalid trade: $trade")
    }
  }
}

File: app/src/main/scala/com/trade/validationService/ValidationService.java
package com.trade.validationService;

import com.crankuptheamps.client.HAClient;
import com.crankuptheamps.client.Message;
import com.crankuptheamps.client.MessageStream;
import com.trade.amps.AmpsClientUtil;

public class ValidationService {
    public static void main(String []args) throws Exception{
        HAClient client = AmpsClientUtil.connect("ValidationService");
        System.out.println("Validation Service connected, waiting for trades...");

        MessageStream ms = client.subscribe("trades.enriched").timeout(0);
        for(Message msg :ms){
            String trade = msg.getData();

            //Simple validation
            boolean valid = trade.contains("qty");
            if(valid){
                client.publish("trades.validated", trade);
                System.out.println("Validated trade: "+trade);
            } else {
                System.out.println("Invalid trade: "+trade);
            }
        }

        client.wait(20000);
        client.close();
    }
}

